### PR TITLE
Optimistically classify the 'await' keyword

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/2301

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/mlangide

**Comment by DustinCampbell** [comment]
The issue mentions VB as well -- is there a corresponding change for that?

**Comment by CyrusNajmabadi** [comment]
Good point.  VB side done :)

It was a bit more interesting since things may be syntactically fine, but we still want 'await' to be classified eagerly as a keyword.  To make tihs possible, i decided to classify things such that if we see 'await' as an identifier expression, and it doesn't bind to anything, then we make it a keyword.

File: src/Workspaces/VisualBasic/Portable/Classification/Classifiers/IdentifierNameSyntaxClassifier.vb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        Public Overrides ReadOnly Property SyntaxNodeTypes As IEnumerable(Of Type)
+            Get
+                Return {GetType(IdentifierNameSyntax)}
+            End Get
+        End Property
+
+        Public Overrides Function ClassifyNode(syntax As SyntaxNode, semanticModel As SemanticModel, cancellationToken As CancellationToken) As IEnumerable(Of ClassifiedSpan)
+            Dim identifierName = DirectCast(syntax, IdentifierNameSyntax)
+            Dim identifier = identifierName.Identifier
+            If StringComparer.OrdinalIgnoreCase.Equals(identifier.ValueText, AwaitText) Then
```
Comment (DustinCampbell) [inline src/Workspaces/VisualBasic/Portable/Classification/Classifiers/IdentifierNameSyntaxClassifier.vb]: I seem to recall there being some helper class for comparing identifiers in VB, but I can't find it. Maybe my memory isn't what it used to be? @jasonmalinowski, do you know?

**Comment by DustinCampbell** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
Are you thinking about ```CaseInsensitiveComparison.Compare```?  That looks pretty good to me to switch to.

**Comment by DustinCampbell** [comment]
Yup, that's the one I was thinking. It has some special processing.

**Comment by CyrusNajmabadi** [comment]
retest this please

File: src/EditorFeatures/CSharpTest/Classification/SyntacticClassifierTests.cs [hunk 2]
```diff
@@ -3605,5 +3605,48 @@ public void LoadDirective()
...
+                Punctuation.CloseCurly);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Classification)]
+        public void CompleteAwaitInNonAsyncContext()
+        {
+            var code = @"
+void M()
+{
+    var x = await;
```
Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/Classification/SyntacticClassifierTests.cs]: What about non-assignment cases:
```
void M()
{
    await
}
```

Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/Classification/SyntacticClassifierTests.cs]: And calling methods:
```
void M(int x)
{
    M(await
}
```

What happens in these cases when brace completion has already put the `)` in?

**Comment by Pilchie** [comment]
Would like some tests for a few more syntactic constructs, but otherwise LGTM.

File: src/EditorFeatures/CSharpTest/Classification/SyntacticClassifierTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Classification/SyntacticClassifierTests.cs]: So in the latter two cases this approach will work.  It won't work for the ```M(await)``` case.  I could go with the approach I took with VB and just try to bind it semantically.  If it fails, I can just assume it had to be a keyword...  I wasn't thinking it had to be so 'perfect'.  I was happy with something that would get it right 99/100 times. :)

Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/Classification/SyntacticClassifierTests.cs]: It's definitely a huge improvement.  @ljw1004 @DustinCampbell what do you think?

**Comment by ljw1004** [comment]
Is there a way I can test this fix? (I don't have any Roslyn enlistment and don't plan to get one...)

FWIW, I think it'd be fine to ALWAYS colorize "await" and "async", at least in VB. If someone chooses to use an identifier with that name, well, they can live with poor colorization...

**Comment by DustinCampbell** [comment]
We still color await correctly when it binds to something though, right?

@ljw1004 -- If you're interested in testing the fix with, you should chat with us offline. There are many people in your hallway that can show you the change. :smile:

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
For VB this should be pretty good.  

For ```await```
1. if the parser already thinks it is a keyword, then we colorize it as a keyword.
2. if the parser think it's an IdentifierName, then we try to bind it.  Only if it binds to something do we colorize it as an identifier.  Otherwise we colorize it as a keyword.

**Comment by tannergooding** [comment]
test eta please

**Comment by CyrusNajmabadi** [comment]
retest eta please

**Comment by CyrusNajmabadi** [comment]
test eta please

**Comment by CyrusNajmabadi** [comment]
test this please

**Comment by CyrusNajmabadi** [comment]
retest mac please

**Comment by CyrusNajmabadi** [comment]
retest mac please

**Comment by CyrusNajmabadi** [comment]
retest this please

**Pull request merged**